src/chunkserver/datastore/file_pool.cpp

```diff
@@ -244,24 +252,197 @@ bool FilePool::CheckValid() {
+                std::min(ntotal - nwrite, (uint64_t)kBytesPerWrite_));
+            if (nbytes < 0) {
+                LOG(ERROR) << "Write file failed: " << chunkpath.c_str();
+                return -1;
+            } else if (fsptr_->Fsync(fd) < 0) {
+                LOG(ERROR) << "Fsync file failed: " << chunkpath.c_str();
+                return -1;
+            }
+
+            cleanThrottle_.Add(false, kBytesPerWrite_);
+            nwrite += nbytes;
+        }
+    }
+
+    std::string targetpath = chunkpath + kCleanChunkSuffix_;
+    ret = fsptr_->Rename(chunkpath.c_str(), targetpath.c_str());
+    if (ret < 0) {
+        LOG(ERROR) << "Rename file failed: " << chunkpath.c_str();
+        return -1;
+    }
+
+    return 0;
+}
+
+bool FilePool::CleaningChunk() {
+    auto popBack = [this](std::vector<uint64_t>& chunks,
+        uint64_t& chunksLeft) -> uint64_t {
+        std::unique_lock<std::mutex> lk(mtx_);
+        if (0 == chunks.size()) {
+            return 0;
+        }
+
+        uint64_t chunkid = chunks.back();
+        chunks.pop_back();
+        chunksLeft--;
+        currentState_.preallocatedChunksLeft--;
+        return chunkid;
+    };
+
+    auto pushBack = [this](std::vector<uint64_t>& chunks,
+        uint64_t chunkid, uint64_t& chunksLeft) {
+        std::unique_lock<std::mutex> lk(mtx_);
+        chunks.push_back(chunkid);
+        chunksLeft++;
+        currentState_.preallocatedChunksLeft++;
+    };
+
+    uint64_t chunkid = popBack(dirtyChunks_, currentState_.dirtyChunksLeft);
+    if (0 == chunkid) {
+        return false;
+    }
+
+    // Fill zero to specify chunk
+    int ret = CleanChunk(chunkid, false);
+    if (ret < 0) {
+        pushBack(dirtyChunks_, chunkid, currentState_.dirtyChunksLeft);
+        return false;
+    }
+
+    LOG(INFO) << "Clean chunk success, chunkid: " << chunkid;
+    pushBack(cleanChunks_, chunkid, currentState_.cleanChunksLeft);
+    return true;
+}
+
+void FilePool::CleanWorker() {
+    auto sleepInterval = kSuccessSleepMsec_;
+    while (cleanSleeper_.wait_for(sleepInterval)) {
+        sleepInterval = CleaningChunk() ? kSuccessSleepMsec_ : kFailSleepMsec_;
+    }
+}
+
+bool FilePool::StartCleaning() {
+    if (poolOpt_.needClean && !cleanAlived_.exchange(true)) {
+        ReadWriteThrottleParams params;
+        params.iopsTotal = ThrottleParams(poolOpt_.iops4clean, 0, 0);
+        cleanThrottle_.UpdateThrottleParams(params);
+
+        cleanThread_ = Thread(&FilePool::CleanWorker, this);
+        LOG(INFO) << "Start clean thread ok.";
+    }
+
+    return true;
+}
+
+bool FilePool::StopCleaning() {
+    if (cleanAlived_.exchange(false)) {
+        LOG(INFO) << "Stop cleaning...";
+        cleanSleeper_.interrupt();
+        cleanThread_.join();
+        LOG(INFO) << "Stop clean thread ok.";
+    }
+
+    return true;
+}
+
+uint64_t FilePool::GetChunk(bool needClean, bool* isCleaned) {
+    auto smartPop = [this](std::vector<uint64_t>& chunks1,
```

Comment:
This lambda a little bit complex. It better to define two member functions, e.g., `Push` and `Pop`, and each one takes one argument to determine whether push/pop from dirty pool or clean pool.

By the way, if the function going to change the argument, it should task a pointer instead of a non-const reference.

Reply:
done